### ISSUE TITLE
Hub entering a "loop of death" when connected to PHL (caused by echoed own MQTT message)

### ISSUE THREAD

**Issue by jacotec** [PetHubLocal/pethublocal#11]
I need some advice from a PHL / Python expert (maybe @plambrechtsen ;-) ) with a very strange issue.

I've added a second cat flap to the hub to secure the path from our screened outside area to the garden with a controllable flap. I did this in the cloud to get a proper generated new pethubconfig.json. I've also updated PHL from 2.0.0 to 2.0.1 in this step. I took a snapshot of the VM running PHL before.

As a global status for the cats does not work in this scenario (possible locations: House, screened area, outside) I need the location per cat flap (what I've desribed in #10 ). So I try to learn that much Python to realize this. I attach my current state (functions.py) here. Finally I plan to do this as a pull request.

With an update it gets the MAC of the updating device, and searches for the Index in the config. If not present, it adds new fields in the pethubconfig.json for that particular pet:

```
Update_Mac: xxxxxxxx   -> The MAC of the device which did the last location update
Idx_0: Outside               -> Last Position for device index 0
Idx_1: Inside                   -> Same for index 1
...
Idx_9: Unknown             -> Same for index 9

```
This is also reflected to MQTT so Home Assistant gets these fields as well.

This works pretty well for my first Python attempt, **BUT**:

After some time (or number of actions, no clue) out of the blue the Hub suddenly looses connection and turns into red flashing ears. After some time, the ears are flashing 2-3 times green, then back to red.
In the same intervals the hub is requesting the certificate (which is answered by PHL). From the logs (Certificate crippled):

```
Aug 30 23:45:16 nodered pethublocal[1754]: INFO:pethublocal:Hub Request serial_number=H009-0333547&mac_address=000004916278FCF1&product_id=1&firmware_version=2.43
Aug 30 23:45:16 nodered pethublocal[1754]: INFO:pethublocal:Hub Credentials Response v02:333547:H009-0333547:::1:pethub/hub/H009-0333547:hub.api.surehub.io:MIIKLwIBAzCCCeUGCSqGSIb3DQEHAaCCCdYEggnSMIIJzjCCBEIGCSqGSIb3DQEHBqCCBDMwggQvAgEAMIIEKAYJKoZIhvcNAQcBMFcGCSqGSIb3DQEFDTBKMCkGCSqGSIb3DQEFDDAcBAgJHxK4lyC8+QICCAAwDAYIKoZIhvcNAgkFADAdBglghkgBZQMEAS_XXXXXXXXXX_q4pfKSBiYxN7U0Y3bjm/KgmSgAtfA6woTb31Cow8uaP0FYGNyd7XHbkv+2uLZ7ez6YDxw90/2GSjTO2KgqIDguDkCY6hmULJKtgTbjRKAxjIHPpTxiMSUwIwYJKoZIhvcNAQkVMRYEFG6azFG0CIgQurOhxWHHun1B7dd3MEEwMTANBglghkgBZQMEAgEFAAQgnkeOQnf1u72esAyty4IjCj2fSz95iowakixPFiqClDQECJRSG1lRR9AgAgIIAA==
Aug 30 23:45:16 nodered pethublocal[1754]: INFO:aiohttp.access:10.30.1.21 "POST /api/credentials HTTP/1.1" 200 3712 "-" "curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3"
```

But the hub is not finally connecting.

It's getting **even more strange**: 

When I stop PHL, move the complete "pethublocal" directory somewhere else and restore my previously copied "pethublocal" folder, running "pip install -U ." and then starting it, it still does the same! Although everything is back to the old, original installation (including the unmodified pethublocal.json).

It looks something "gets poisoned" on the machine and I have no clue what. All config and run time data is in "pethublocal.json" which is also restored back to the original.

I've restarted the local MQTT, deleted the "pethub" topic before, rebooted the whole VM - nothing helps.

**The only way to get it back working is to restore my previously saved backup of the whole VM**. Once this is done, the hub reconnects immediately (so it's nothing in the hub).

Any idea in this complete mystery? Especially an explanantion how the Ubuntu VM can get into a "state" where the Hub can't connect and only a snapshot restore helps?

**Comment by jacotec** [comment]
Current version of my "functions.zip"
[functions.zip](https://github.com/PetHubLocal/pethublocal/files/9460823/functions.zip)

**Comment by jacotec** [comment]
Update: I've reworked my change to a much more simple way (Just added "Update_MAC" and "Time" to the pet, so HomeAssistant can track the position via automations).

But the same issue still persists. After hours of working state a simple hub restart and it did not reconnect. It really looks like whatever dies might have to to with the certificate exchange. However, as usual restoring the pethubconfig.json does not help. Also not restoring the full pethublocal directory, again I need to revert the snapshot.

Old snapshot with the config file from the dead state runs and connects (until it dies again).

I have a full logfile from start up to the state where it does not reconnect. Can send it via PM (don't want to expose it to the public due to certs etc.)

As I've changed three things:
- Update from 2.0.0 to 2.0.1
- Added a second cat flap
- Did my extension

I try to revert one by one to see if I find out what specific change caused it.

Still this "strange state" is a full mystery.

**Comment by jacotec** [comment]
Update:

The issue is caused by MQTT, it looks like a "message of death". In the blocked state I've stopped Mosquitto, removed the persistent data, restarted Mosquitto and the issue is gone (until it happens again).

Here's the difference from the view of MQTT debug logs.

**Hub boot / Non-working state**

```
1662013095: New connection from 10.30.1.21:1258 on port 8883.
1662013097: New client connected from 10.30.1.21:1258 as H009-0333547 (p2, c0, k90, u'?SDK=C&Version=3.0.1').
1662013097: Will message specified (20 bytes) (r0, q1).
1662013097: 	pethub/hub/H009-0333547/messages
1662013097: Sending CONNACK to H009-0333547 (1, 0)
1662013097: Sending PUBLISH to H009-0333547 (d1, q1, r0, m12032, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (50 bytes))
1662013097: Sending PUBLISH to H009-0333547 (d1, q1, r0, m12033, 'pethub/hub/H009-0333547/messages', ... (20 bytes))
1662013133: Sending PINGREQ to local.nodered.homeassistant
```


**Hub boot / Working state**

```
1662014024: New connection from 10.30.1.21:1497 on port 8883.
1662014026: New client connected from 10.30.1.21:1497 as H009-0333547 (p2, c0, k90, u'?SDK=C&Version=3.0.1').
1662014026: Will message specified (20 bytes) (r0, q1).
1662014026: 	pethub/hub/H009-0333547/messages
1662014026: Sending CONNACK to H009-0333547 (0, 0)
1662014027: Received SUBSCRIBE from H009-0333547
1662014027: 	pethub/hub/H009-0333547/# (QoS 1)
1662014027: H009-0333547 1 pethub/hub/H009-0333547/#
1662014027: Sending SUBACK to H009-0333547
1662014027: Received PUBLISH from H009-0333547 (d0, q1, r0, m3, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
1662014027: Sending PUBLISH to H009-0333547 (d0, q1, r0, m1, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
1662014027: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
1662014027: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
1662014027: Sending PUBACK to H009-0333547 (m3, rc0)
1662014027: Received PUBLISH from PetHubLocal (d0, q0, r1, m0, 'pethub/ha/H009-0333547_Hub/state', ... (221 bytes))
1662014027: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r1, m0, 'pethub/ha/H009-0333547_Hub/state', ... (221 bytes))
1662014027: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/ha/H009-0333547_Hub/state', ... (221 bytes))
1662014027: Received PUBACK from H009-0333547 (Mid: 1, RC:0)
```

First, I thought it would be this fact:
In the non-working state Mosquitto sends two PUBLISH messages right one after the other to the hub directly after it did connect. Then the hub just does nothing, it leaves the TCP connection open and goes red. After around 1-2 minutes, we see a new certificate request, the hub connects again to Mosquitto (Mosquitto mourning that the connection anready exists and that it's closing the old one) ... gets the same two PUBLISH messages immediately and the show continues in a loop.

In the working state we have a clear CONNACK - SUBSCRIBE and then one single PUBLISH to the hub.

I have another working log where only one PUBLISH is done before the SUBSCRIBE:

```
[09:46:44]: New connection from 10.30.1.21:1501 on port 8883.
[09:46:46]: New client connected from 10.30.1.21:1501 as H009-0333547 (p2, c0, k90, u'?SDK=C&Version=3.0.1').
[09:46:46]: Will message specified (20 bytes) (r0, q1).
[09:46:46]: 	pethub/hub/H009-0333547/messages
[09:46:46]: Sending CONNACK to H009-0333547 (1, 0)
[09:46:46]: Sending PUBLISH to H009-0333547 (d1, q1, r0, m11927, 'pethub/hub/H009-0333547/messages', ... (20 bytes))
[09:46:47]: Received SUBSCRIBE from H009-0333547
[09:46:47]: 	pethub/hub/H009-0333547/# (QoS 1)
[09:46:47]: H009-0333547 1 pethub/hub/H009-0333547/#
[09:46:47]: Sending SUBACK to H009-0333547
[09:46:47]: Received PUBACK from H009-0333547 (Mid: 11927, RC:0)
[09:46:47]: Received PUBLISH from H009-0333547 (d0, q1, r0, m3, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
[09:46:47]: Sending PUBLISH to H009-0333547 (d0, q1, r0, m11928, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
[09:46:47]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (37 bytes))
```

So, the hub does not generally seem to have problems with getting a PUBLISH before the SUBSCRIBE, it looked like the hub crashes or gets into a strange state when it gets two PUBLISH messages right one after the other before it has acknowledged.

Unfortunately that was not the case, because I've then configured

`max_inflight_messages 1`

to my mosquitto.conf

This should prevent Mosquitto from publish more than one message as the same time.

Unfortunately half an hour later the hub was red again.

These are the last lines in the Mosquitto log for the hub:

```
[11:15:07]: Received PUBLISH from H009-0333547 (d0, q1, r0, m34, 'pethub/hub/H009-0333547/messages/117503CFF9D5B370', ... (34 bytes))
[11:15:07]: Sending PUBLISH to H009-0333547 (d0, q1, r0, m12029, 'pethub/hub/H009-0333547/messages/117503CFF9D5B370', ... (34 bytes))
[11:15:07]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages/117503CFF9D5B370', ... (34 bytes))
[11:15:07]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages/117503CFF9D5B370', ... (34 bytes))
[11:15:07]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages/117503CFF9D5B370', ... (34 bytes))
[11:15:07]: Sending PUBACK to H009-0333547 (m34, rc0)
[11:15:07]: Received PUBACK from H009-0333547 (Mid: 12029, RC:0)
[11:15:07]: Received PUBLISH from H009-0333547 (d0, q1, r0, m35, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:15:07]: Sending PUBLISH to H009-0333547 (d0, q1, r0, m12030, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:15:07]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:15:07]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:15:07]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:15:07]: Sending PUBACK to H009-0333547 (m35, rc0)
[11:15:07]: Received PUBLISH from PetHubLocal (d0, q0, r1, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:15:07]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r1, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:15:07]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:15:07]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:15:07]: Received PUBLISH from PetHubLocal (d0, q0, r1, m0, 'pethub/ha/H009-0333547_117503CFF9D5B370/state', ... (134 bytes))
[11:15:07]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r1, m0, 'pethub/ha/H009-0333547_117503CFF9D5B370/state', ... (134 bytes))
[11:15:07]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/ha/H009-0333547_117503CFF9D5B370/state', ... (134 bytes))
[11:15:07]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/ha/H009-0333547_117503CFF9D5B370/state', ... (134 bytes))
[11:15:07]: Received PUBACK from H009-0333547 (Mid: 12030, RC:0)

[11:15:30]: Received PINGREQ from H009-0333547
[11:15:30]: Sending PINGRESP to H009-0333547

[11:16:06]: Received PUBLISH from H009-0333547 (d0, q1, r0, m36, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (34 bytes))
[11:16:06]: Sending PUBLISH to H009-0333547 (d0, q1, r0, m12031, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (34 bytes))
[11:16:06]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (34 bytes))
[11:16:06]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (34 bytes))
[11:16:06]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (34 bytes))
[11:16:06]: Sending PUBACK to H009-0333547 (m36, rc0)
[11:16:06]: Received PUBACK from H009-0333547 (Mid: 12031, RC:0)
[11:16:06]: Received PUBLISH from H009-0333547 (d0, q1, r0, m37, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:16:06]: Sending PUBLISH to H009-0333547 (d0, q1, r0, m12032, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:16:06]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:16:06]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:16:06]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
[11:16:06]: Sending PUBACK to H009-0333547 (m37, rc0)
[11:16:06]: Received PUBLISH from PetHubLocal (d0, q0, r1, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:16:06]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r1, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:16:06]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:16:06]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/ha/H009-0333547_Hub/state', ... (220 bytes))
[11:16:06]: Received PUBLISH from PetHubLocal (d0, q0, r1, m0, 'pethub/ha/H009-0333547_651B02CFF9D5B370/state', ... (134 bytes))
[11:16:06]: Sending PUBLISH to local.nodered.homeassistant (d0, q0, r1, m0, 'pethub/ha/H009-0333547_651B02CFF9D5B370/state', ... (134 bytes))
[11:16:06]: Sending PUBLISH to mqtt-explorer-79e9c8e8 (d0, q0, r0, m0, 'pethub/ha/H009-0333547_651B02CFF9D5B370/state', ... (134 bytes))
[11:16:06]: Sending PUBLISH to PetHubLocal (d0, q0, r0, m0, 'pethub/ha/H009-0333547_651B02CFF9D5B370/state', ... (134 bytes))

[11:17:29]: New connection from 10.30.1.21:1412 on port 8883.
[11:17:30]: Client H009-0333547 already connected, closing old connection.
[11:17:30]: New client connected from 10.30.1.21:1412 as H009-0333547 (p2, c0, k90, u'?SDK=C&Version=3.0.1').
[11:17:30]: Will message specified (20 bytes) (r0, q1).
[11:17:30]: 	pethub/hub/H009-0333547/messages
[11:17:30]: Sending CONNACK to H009-0333547 (1, 0)
[11:17:30]: Sending PUBLISH to H009-0333547 (d1, q1, r0, m12032, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
... ** hub crashed **

```

So, after receiving the echoed own 40 bytes message from 11:06:06 (m12032), there was no single life sign from the hub. The hub did no MQTT pings, nothing. Instead, it's again in the "loop of death", as with a new start Mosquitto publishes this "40 byte message of death" again to the hub, it dies ... and so on.

Below is the content of these last two messages from the hub:

Last messages content:
Received PUBLISH from H009-0333547 (d0, q1, r0, m36, 'pethub/hub/H009-0333547/messages/651B02CFF9D5B370', ... (34 bytes))
63107854 0210 132 32 33 3 aa 09 10

[11:16:06]: Received PUBLISH from H009-0333547 (d0, q1, r0, m37, 'pethub/hub/H009-0333547/messages', ... (40 bytes))
**63107855 0220 132 33 71 5 1b 82 34 d4 8e**

Stopping Mosquitto, removing the database and restarting and all starts again. Until the next "bad message".

I'm puzzled ...

**Comment by jacotec** [comment]
I've reconnected the hub to the cloud, played a bit with the settings of the last added flap (add curfew, removed curfew) and left the hub for 30 mins connected to the cloud.

Re-generated the PHL config afterwards and now also the curfew entries are generated. The hub did not die in the last 2 days.

So I think something was screwed in the hub, it seems to have a local stored non-volatile config as well.

@plambrechtsen Should I leave this open as this would be worth for a FAQ if that happens to someone else, too? It costs me two days and quite some logging and analyzing and has potential to distress a less experienced user? If not, feel free to close.

**Comment by jacotec** [comment]
@plambrechtsen After more than 2 weeks my hub crashed again into red today with the same "message of death".
I'm pretty puzzled why this does not happen when it's on its original cloud (as this should have the same MQTT mechanism).
Do you have any theory?

**Comment by darkorb** [comment]
> @plambrechtsen Should I leave this open as this would be worth for a FAQ if that happens to someone else, too? It costs me two days and quite some logging and analyzing and has potential to distress a less experienced user? If not, feel free to close.

For what it's worth, I have ran into the same issue documented here. Purging the persisted data in MQTT resolved the issue instantly (but caused others heh). Thanks for taking the time to document it.

**Comment by jacotec** [comment]
It'll be not easy to debug this as this is a normal MQTT behavior at this point and it happens before PHL is in the game.

However, it happens not with the cloud which is strange.

@plambrechtsen has documented that he wonders that the cloud send every message back to the hub. I think what he saw was the normal MQTT QOS1 behavior, so the cloud should do the same. The question is: Why does the hub not send this "message of death" to the cloud - or why does the cloud not echo it back to the hub. I can't imagine this is an intentional kind of protection with the firmware we use.

**Comment by jacotec** [comment]
> > @plambrechtsen Should I leave this open as this would be worth for a FAQ if that happens to someone else, too? It costs me two days and quite some logging and analyzing and has potential to distress a less experienced user? If not, feel free to close.
> 
> For what it's worth, I have ran into the same issue documented here. Purging the persisted data in MQTT resolved the issue instantly (but caused others heh). Thanks for taking the time to document it.

This really drives me mad every few weeks. Especially as only the sysadmin (me ;-) ) can solve this situation. This causes a bad WAF.

I've thought about a concept to automate this (see #17) and hope someone is able to code this. Should be easy for an experienced Python programmer.

I really have no idea how the original SurePetcare cloud avoids this. From the MQTT specification the same should happen there as well ... except they have implemented a workaround in their MQTT broker.

**Comment by jacotec** [comment]
Hi @olivier-m 

> I implemented another workaround by simply not using the persistence at all. So far, it's been working smoothly without any issue. So maybe it's only a matter of changing the documentation of mosquitto's configuration ; remove `persistence_location` and add `persistence false`.

That's a good approach, too. But how does the system behave when you restart Mosquitto on the PHL installation? Will be current settings (locks etc.) be lost in that case?

**Comment by jacotec** [comment]
As this is a Hub firmware issue, it can't be really fixed at the root.
I've implemented a workaround in PR #22 which at least should solve the issue automatically latest around one hour after the hub crashed. I don't see any other way to react faster, but it'll prevent the "sysadmin" from fixing it manually every time it happens.

@plambrechtsen Please kindly review my PR

**Comment by plambrechtsen** [comment]
I am re-opening this issue and apologise I haven't been responsive (at all) for months now, long involved story with that one. I am **really** interested in this problem as I haven't had it myself, but I have a mqtt managed power switch with the USB Power supply and hub plugged into it, and reboot the hub weekly as I had weird issues where the hub would just drop off and found that was the least stress to get it back online. My only issue with nuking the persistence DB and restarting Mosquitto is it won't really work for folks using HAOS.
I'm going to connect up my spare hub, and a spare door I have to see if there is anything that can be done.
To me  the pragmatic way to solve it is using the watchdog approach you wrote, but listen out for two things.
- Wait for the hourly uptime messages
- If the MQTT Last Will message comes through
Then have the option to restart MQTT, or do what I am doing and drop a MQTT message into the tasmota topic to power cycle the port.

Another thing I am wondering about is if this is fixed in the newer firmware. I have managed to byte patch my own certificate and hostname into the 2.201 firmware as per https://github.com/PetHubLocal/pethublocal/tree/firmware201/pethublocal/firmware
So.... I wonder if there is newer firmware from surepet that might fix this as IMHO nuking MQTT or rebooting the hub is a pretty brutal way to sort it.

**Comment by jacotec** [comment]
Hi @plambrechtsen ,
I fully agree that in any case the root cause should be fixed. But with the fact that the "message of death" is originating from the hub itself and specified MQTT behavior leads to the race condition we face, I saw no other solution (for now) to get around it with a watchdog.

When the issue occurs, the only way to fix it is to stop Mosquitto, delete persistent data and restart it. Until this is done you have no control over the connected devices. In that situation my wife can't lock the cats into the screened area without manually locking the flaps and we have no remote control without me SSHing into the PHL VM solving the issue.

I'd propose to merge the PR as a first responder action because others are affected as well. The log entries will tell if the situation occurs or not.

Interesting question: Why is the Sure Petcare MQTT not getting into this situation as the QOS spec normally would define the same behavior as Mosquitto is showing.

I can try the newer firmware if you can show me a way to update without risk. I just have my production installation, no playground available.

PS: Restarting your hub weekly just "to be sure" is not less brutal than my watchdog which acts only if we're into the bad situation ;-)

**Comment by Flachzange** [comment]
Thanks a lot @jacotec for digging into this issue and for finding the workaround solution. After more than 6 weeks, my hub stopped working as well at the weekend and I was puzzled too as I was not even at home at that point in time. (usually I break the stuff myself). Unfortunately, the watchdog does not work for me as my mosquitto is running in another container also used for other applications.

**Comment by flyize** [comment]
Could you use the self-contained Mosquitto in the pethublocal Docker container?

**Comment by jacotec** [comment]
@Flachzange I agree to @flyize, I also have a "main MQTT broker" running in a different VM which serves my HomeAssistant - the Mosquitto in the PetHubLocal VM just does the handling of the Petcare stuff and connects upwards to my master MQTT.

Also restarting and deleting the persistent database of the "big main Mosquitto" would not be something what I'd like. So maybe you can use the local Mosquitto with PHL as a slave and just restart this?

Basically my watcdog works well. Last week, sitting in the living room my hub turned red. As I did not need to control a flap at this point I just waited ... approximately one hour later the situation was cleared by the watchdog and the hub went online again.

**Comment by Flachzange** [comment]
I understand your points, but if I am not mistaken, the pethublocal docker container does not ship mosquitto already which would actually let me configure and using the local mosquitto or did I overlook this?

**Comment by flyize** [comment]
It does. The reason is that HA MQTT requires auth, and the pethub can't use auth. So it forwards the messages to the HA MQTT.

**Comment by Flachzange** [comment]
> It does. The reason is that HA MQTT requires auth, and the pethub can't use auth. So it forwards the messages to the HA MQTT.

Honestly, I don't think that is true. The Dockerfile in this github repo does not contain an MQTT Server. It is something that needs to be setup additionally, but then the watchdog stuff isn't working anymore. Only alternative I see is, to put this onto a dedicated VM (which is actually a bit too much if you don't need it for something else anyhow)

**Comment by flyize** [comment]
I stand corrected, the MQTT server is in the Addon

` THis is because the core Mosquitto broker requires authentication which is an issue as the Hub doesn't support MQTT authentication. So the addon includes a Mosquitto broker and connects to the core broker as a remote broker with a username and password.`

**Comment by Flachzange** [comment]
Ah, see, we are talking different topics here. I don't use neiher the addon nor home assistant. But still, the watchdog included in pethublocal restarts a local mosquitto. Even the one in the addon is not accessible by the watchdog. At least that would be my understanding.

**Comment by plambrechtsen** [comment]
Closing this issue as I now have a whole new approach. This python code is great for a reference but I won't be making any changes to it going forward. If you are interested in my new approach send me an email pethublocal at gmail . com.
